parcels/kernels/advection.py

```diff
@@ -100,3 +101,159 @@ def AdvectionRK45(particle, fieldset, time):
+    if grid.gtype < 2:
+        px = np.array([grid.lon[xi], grid.lon[xi + 1], grid.lon[xi + 1], grid.lon[xi]])
+        py = np.array([grid.lat[yi], grid.lat[yi], grid.lat[yi + 1], grid.lat[yi + 1]])
+    else:
+        px = np.array([grid.lon[yi, xi], grid.lon[yi, xi + 1], grid.lon[yi + 1, xi + 1], grid.lon[yi + 1, xi]])
+        py = np.array([grid.lat[yi, xi], grid.lat[yi, xi + 1], grid.lat[yi + 1, xi + 1], grid.lat[yi + 1, xi]])
+    if grid.mesh == 'spherical':
+        px[0] = px[0]+360 if px[0] < particle.lon-225 else px[0]
+        px[0] = px[0]-360 if px[0] > particle.lat+225 else px[0]
+        px[1:] = np.where(px[1:] - px[0] > 180, px[1:]-360, px[1:])
+        px[1:] = np.where(-px[1:] + px[0] > 180, px[1:]+360, px[1:])
+    if withW:
+        pz = np.array([grid.depth[zi], grid.depth[zi+1]])
+        dz = pz[1] - pz[0]
+    else:
+        dz = 1.
+
+    c1 = fieldset.UV.dist(px[0], px[1], py[0], py[1], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 0.), py))
+    c2 = fieldset.UV.dist(px[1], px[2], py[1], py[2], grid.mesh, np.dot(i_u.phi2D_lin(1., eta), py))
+    c3 = fieldset.UV.dist(px[2], px[3], py[2], py[3], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 1.), py))
+    c4 = fieldset.UV.dist(px[3], px[0], py[3], py[0], grid.mesh, np.dot(i_u.phi2D_lin(0., eta), py))
+    dx = (c4 + c2)/2.  # TODO is this the best way to compute dx and dy?
+    dy = (c1 + c3)/2.
+
+    if withW:
+        U0 = direction * fieldset.U.data[ti, zi+1, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, zi+1, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, zi+1, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, zi+1, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi+1, xi+1] * c3 * dz
+    else:
+        U0 = direction * fieldset.U.data[ti, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi+1, xi+1] * c3 * dz
+
+    def compute_ds(F0, F1, r, direction, tol):
+        up = F0 * (1-r) + F1 * r
+        r_target = 1. if direction * up >= 0. else 0.
+        B = F0 - F1
+        delta = - F0 - B * 0.
```

Comment:
ehm, that statement I don't fully understand (or: the intent of it) - isn't that just `delta = -F0` ? Unless, of course, we are dealing with values close to the numerical accuracy, where `delta` could switch signs (iff F0 ~ +0, B ~ -0 and abs(B) > abs(F0))

Reply:
I left the `B * 0.` in to stay close to the notation of Döös et al (2017); [their equation (17)](https://www.geosci-model-dev.net/10/1733/2017/gmd-10-1733-2017.pdf) where in our unit cell by definition r<sub>i-1</sub>=0.

But I agree it could only cause numerical errors, so will remove